nbd/src/main.cpp

```diff
@@ -95,6 +97,112 @@ static void Usage() {
+                                            (fs.rdbuf())->fd();
+    if (0 == flock(fd, LOCK_SH )) {
+        // check if the record exists
+        std::string lineString;
+        while (getline(fs, lineString)) {
+            if (!lineString.empty() && lineString[0] != '#') {
+                std::vector<std::string> lineArray;
+                std::stringstream ss(lineString);
+                std::string str;
+                while (getline(ss, str, '\t')) {
+                    lineArray.push_back(str);
+                }
+                if (lineArray.size() != 2 && lineArray.size() != 3) {
+                    std::cerr << "curve-nbd: curvetab file content error.";
+                    fs.close();
+                    return;
+                }
+                if (lineArray[0] == nbdConfig->imgname) {
+                    fs.close();
+                    return;
+                }
+            }
+        }
+        fs.close();
+        flock(fd, LOCK_UN);
+    }
+    // write new record
+    fs.open(CURVETAB_PATH, std::fstream::app);
+    if (!fs.is_open()) {
+        std::cerr << "curve-nbd: open curvetab file failed.";
+        return;
+    }
+    fd = static_cast<__gnu_cxx::stdio_filebuf< char > * const>
+                                           (fs.rdbuf())->fd();
+    if (0 == flock(fd, LOCK_EX)) {
+        fs << nbdConfig->imgname << '\t' << nbdConfig->devpath << '\n';
+        fs.close();
+        flock(fd, LOCK_UN);
+    }
+}
+
+// delete record when unmap
+static void DelRecord() {
+    std::fstream fs;
+    int fd;
+    fs.open(CURVETAB_PATH, std::fstream::in);
+    if (!fs.is_open()) {
+        std::cerr << "curve-nbd: open curvetab file failed.";
+        return;
+    }
+    fd = static_cast<__gnu_cxx::stdio_filebuf< char > * const>
+                                           (fs.rdbuf())->fd();
+    std::string lineString;
+    std::string newContent;
+    if (0 == flock(fd, LOCK_SH)) {
+        // find record need deleted
+        while (getline(fs, lineString)) {
+            if (!lineString.empty() && lineString[0] != '#') {
+                std::vector<std::string> lineArray;
+                std::stringstream ss(lineString);
+                std::string str;
+                while (getline(ss, str, '\t')) {
+                    lineArray.push_back(str);
+                }
+                if (lineArray.size() != 2 && lineArray.size() != 3) {
+                    std::cerr << "curve-nbd: curvetab file content error.";
+                    fs.close();
+                    return;
+                }
+                if (lineArray[1] != nbdConfig->devpath) {
+                    newContent += lineString;
+                    newContent += "\n";
+                }
+            } else {
+                newContent += lineString;
+                newContent += "\n";
+            }
+        }
+        fs.close();
+        flock(fd, LOCK_UN);
+    }
+    // update record info
```

Comment:
Note there is time window in the read and write logic.

Reply:
> Note there is time window in the read and write logic.

what is mean?